clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -105,7 +105,9 @@ public AbstractConfig(ConfigDef definition, Map<?, ?> originals,  Map<String, ?>
                 throw new ConfigException(entry.getKey().toString(), entry.getValue(), "Key must be a string.");
 
         this.originals = resolveConfigVariables(configProviderProps, (Map<String, Object>) originals);
-        this.values = definition.parse(this.originals);
+        // pass a copy to definition.parse. Otherwise, the definition.parse adds all keys of definitions to "used" group
+        // since definition.parse needs to call "RecordingMap#get" when checking all definitions.
+        this.values = definition.parse(new HashMap<>(this.originals));
```

Comment:
> I tried running console-producer with/without this PR. It doesn't seem to WARN any unused SSL configs in either test. Do you know why?

@junrao this is the root cause.

Reply:
ConfigDef#parse (https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java#L468) tries to get all elements from input maps so all gets are recorded. In order to avoid recording, we pass a copy instead of RecordingMap.

Reply:
> However, that code seems no longer necessary since we are now setting clientId in ProducerConfig.postProcessParsedConfig(). Could we just avoid constructing ProducerConfig there?

I don't think so. The configs passed to configurable object is ```origins``` so the generated "client.id" is not included. However, your feedback inspires me that we don't need to create a new ```ProducerConfig```. Instead, we can use overrideConfig to set generated client.id to those configurable object. Will update it later.

Reply:
> But it seems this is still needed? 

It is not necessary with the latest change. I kept it as a total solution (if someone pass ```RecordingMap``` in the future). However, I'm going to remove it to make this PR simpler.